patroni/validator.py

```diff
@@ -0,0 +1,228 @@
+#!/usr/bin/env python3
+import logging
+import os
+import socket
+
+from patroni.utils import split_host_port
+from patroni.ctl import find_executable
+
+logger = logging.getLogger(__name__)
+
+
+def data_directory_empty(data_dir):
+    if os.path.isfile(os.path.join(data_dir, "global", "pg_control")):
+        return False
+    if not os.path.exists(data_dir):
+        return True
+    return all(
+        os.name != "nt" and (n.startswith(".") or n == "lost+found")
+        for n in os.listdir(data_dir)
+    )
+
+
+def config_validator(config):
```

Comment:
The whole function is huge, it would make sense to break it into multiple smaller functions.
It will simplify future maintenance and testing.

Reply:
Rewritten into Schema class and separate validators